### PR TITLE
BUG: Stack/unstack do not return subclassed objects (GH15563)

### LINKED ISSUE #15563: Preserve subclass family on reshape operations
Subclassed Series and DataFrames with custom `_constructor`, `_constructor_sliced`, and `_constructor_expanddim` methods return pandas Series and DataFrame objects on reshape operations:

```python
>>> import numpy as np, pandas as pd
>>> import pandas.util.testing as tm
>>> df = tm.SubclassedDataFrame(np.random.random((4,5)))
>>> df
          0         1         2         3         4
0  0.151944  0.948561  0.639122  0.718071  0.296193
1  0.004606  0.978378  0.827614  0.479320  0.495945
2  0.158874  0.535476  0.173194  0.065292  0.365851
3  0.772392  0.105038  0.671064  0.164165  0.795803
>>> type(df)
<class 'pandas.util.testing.SubclassedDataFrame'>
```

Slicing operations use `_constructor_sliced`:
```python
>>> type(df[0])
<class 'pandas.util.testing.SubclassedSeries'>
```

Reshape operations return regular pandas objects:
```python
>>> type(df.stack())
<class 'pandas.core.series.Series'>
```

It would be great if this returned a `SublcassedSeries`. Same goes for unstack and pivot.

### PR DESCRIPTION
**Pull request by rockg**
- [x] closes #15563
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Basically picked from #15655 to support stack/unstack in subclassed DataFrames and Series.

### CODE REVIEW

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -345,7 +345,8 @@ Reshaping
 - Fixed construction of a :class:`Series` from a ``dict`` containing ``NaN`` as key (:issue:`18480`)
 - Bug in :func:`Series.rank` where ``Series`` containing ``NaT`` modifies the ``Series`` inplace (:issue:`18521`)
 - Bug in :func:`Dataframe.pivot_table` which fails when the ``aggfunc`` arg is of type string.  The behavior is now consistent with other methods like ``agg`` and ``apply`` (:issue:`18713`)
-
+- Bug in :func:`DataFrame.stack`, `DataFrame.unstack`, `Series.unstack` which were not returning subclasses (:issue:`15563`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: need to have the ``:func:`` for each one

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -37,8 +37,23 @@ class _Unstacker(object):
...
         Level to "unstack". Accepts a name for the level.
+    value_columns : Index, optional
+        Pandas ``Index`` or ``MultiIndex`` object if unstacking a DataFrame
+    fill_value : scalar, optional
+        Default value to fill in missing values if subgroups do not have the
+        same set of labels. By default, missing values will be replaced with
+        the default fill value for that data type, NaN for float, NaT for
+        datetimelike, etc. For integer types, by default data will converted to
+        float and missing values will be set to NaN.
+    constructor : object, default DataFrame
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: thanks!

File: pandas/core/reshape/reshape.py [hunk 3]
```diff
@@ -85,6 +100,13 @@ def __init__(self, values, index, level=-1, value_columns=None,
         self.values = values
         self.value_columns = value_columns
         self.fill_value = fill_value
+        if constructor is None:
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: hmm I hate if then's like this. any better way?

File: pandas/core/reshape/reshape.py [hunk 4]
```diff
@@ -476,12 +499,13 @@ def _unstack_frame(obj, level, fill_value=None):
         unstacker = partial(_Unstacker, index=obj.index,
                             level=level, fill_value=fill_value)
         blocks = obj._data.unstack(unstacker)
-        klass = type(obj)
+        klass = obj._constructor
         return klass(blocks)
     else:
         unstacker = _Unstacker(obj.values, obj.index, level=level,
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: might be a slightly bigger change, but could pass ``obj`` to ``_Unstacker`` directly and *might* be simpler.

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good.

**Comment by rockg** [comment]
Tried to incorporate obj into `_Unstacker`.  Take a look.  When there's a mixed frame being unstacked the underlying blocks are unstacked and so passing obj doesn't really work.

**Comment by jreback** [comment]
can you revert back to what you had. that was ok for now. its not as clear this way.

**Comment by rockg** [comment]
All green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rockg

**Pull request merged**